### ISSUE TITLE
Backward incompatible changes to cron locking mechanism

### ISSUE THREAD

**Issue by jampolm** [magento/magento2#29924]
### Summary (*)

In 2.3.5 next change was done that broke backward compatibility
MC-22819: Fix deadlocks related to crons / consumers
https://github.com/magento/magento2/commit/ce5b6f6e73b2ebe16949d6f713f230b02119e4f8#diff-fcb7a0e44160e6abfceb876e72be5052L241

The locking mechanism on cron job level was introduced and, for some reason, locking mechanism on cron group level was abandoned. Cron group lock is still working for 
$this->cleanupJobs($groupId, $currentTime);
$this->generateSchedules($groupId);
But $this->processPendingJobs($groupId, $jobsRoot, $currentTime); was moved out from cron group lock.
Prior to this change cron jobs from the same cron group couldn’t run simultaneously, now it is possible.

### Examples (*)

Breaking change:
https://github.com/magento/magento2/commit/ce5b6f6e73b2ebe16949d6f713f230b02119e4f8#diff-fcb7a0e44160e6abfceb876e72be5052L241
$this->processPendingJobs($groupId, $jobsRoot, $currentTime); was moved out from cron group lock.

### Proposed solution

I would keep both, locking on group and job level. However, due to the limitation of nested naming locks in MySql 5.6, I was forced to roll-back to the previous pre 2.3.5 locking mechanism. But for 5.7 and higher, it should be pretty easy to implement locking on both levels.

### Information on your environment
Magento 2.3.5 p1

### Steps to reproduce
1)Mark all indexers as invalid
2)run `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1 &`
3)Check that indexer_reindex_all_invalid is running
`SELECT * FROM cron_schedule WHERE (`job_code` = 'indexer_reindex_all_invalid') AND (`status` = 'running')`
3)while indexer_reindex_all_invalid is running call one more `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1 &`

### Expected and actual results
**Expected**: Second bin/magento cron:run will do nothing and the next entry will appear in cron.log
[2020-09-07 14:12:54] report.WARNING: Could not acquire lock for cron group: index, skipping run [] []
**Actual**: Cron jobs indexer_reindex_all_invalid and indexer_update_all_views run in parallel.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by jampolm** [comment]
@magento I am working on this

**Comment by igorwulff** [comment]
We are seeing issues on a live Magento 2.4.3 installation where indexer_reindex_all_invalid and indexer_update_all_views are running in parallel (or so it seems from our cronjob manager) and causing issues for the catalog_category_product indexers. Since the code logic can't handle multiple processes running it in parallel properly. This sometimes causes empty categories on the production environment, damaging sales. Do you know if this issue described in this ticket is still an problem in later Magento releases?

**Comment by kamephis** [comment]
Hi @igorwulff have you already found the reason for your issue. We are experiencing the same on 2.4.3-p1 (live).

**Comment by igorwulff** [comment]
After long conversations and demonstrating the issues, we received a patch which seemed to have fixed the issue. It was given the code MDVA-405550, and I don't know if this patch is was shared elsewhere. 

NOTE: I do not take any responsibility for this patch, and this needs to be tested thoroughly case by case.
```
diff --git a/vendor/magento/module-catalog/Model/Indexer/Category/Product.php b/vendor/magento/module-catalog/Model/Indexer/Category/Product.php
index f5a8c33cfa6..89a6e6d3c25 100644
--- a/vendor/magento/module-catalog/Model/Indexer/Category/Product.php
+++ b/vendor/magento/module-catalog/Model/Indexer/Category/Product.php
@@ -5,7 +5,10 @@
  */
 namespace Magento\Catalog\Model\Indexer\Category;

+use Magento\Framework\App\ObjectManager;
 use Magento\Framework\Indexer\CacheContext;
+use Magento\Framework\Indexer\IndexMutexException;
+use Magento\Framework\Indexer\IndexMutexInterface;

 /**
  * Category product indexer
@@ -41,19 +44,27 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      */
     protected $cacheContext;

+    /**
+     * @var IndexMutexInterface
+     */
+    private $indexMutex;
+
     /**
      * @param Product\Action\FullFactory $fullActionFactory
      * @param Product\Action\RowsFactory $rowsActionFactory
      * @param \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry
+     * @param IndexMutexInterface|null $indexMutex
      */
     public function __construct(
         Product\Action\FullFactory $fullActionFactory,
         Product\Action\RowsFactory $rowsActionFactory,
-        \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry
+        \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry,
+        ?IndexMutexInterface $indexMutex = null
     ) {
         $this->fullActionFactory = $fullActionFactory;
         $this->rowsActionFactory = $rowsActionFactory;
         $this->indexerRegistry = $indexerRegistry;
+        $this->indexMutex = $indexMutex ?? ObjectManager::getInstance()->get(IndexMutexInterface::class);
     }

     /**
@@ -61,6 +72,7 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      *
      * @param int[] $ids
      * @return void
+     * @throws IndexMutexException
      */
     public function execute($ids)
     {
@@ -84,11 +96,17 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      * Execute full indexation
      *
      * @return void
+     * @throws IndexMutexException
      */
     public function executeFull()
     {
-        $this->fullActionFactory->create()->execute();
-        $this->registerTags();
+        $this->indexMutex->execute(
+            static::INDEXER_ID,
+            function () {
+                $this->fullActionFactory->create()->execute();
+                $this->registerTags();
+            }
+        );
     }

     /**
@@ -107,6 +125,7 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      *
      * @param int[] $ids
      * @return void
+     * @throws IndexMutexException
      */
     public function executeList(array $ids)
     {
@@ -118,6 +137,7 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      *
      * @param int $id
      * @return void
+     * @throws IndexMutexException
      */
     public function executeRow($id)
     {
@@ -129,18 +149,22 @@ class Product implements \Magento\Framework\Indexer\ActionInterface, \Magento\Fr
      *
      * @param int[] $ids
      * @return $this
+     * @throws IndexMutexException
      */
     protected function executeAction($ids)
     {
         $ids = array_unique($ids);
         $indexer = $this->indexerRegistry->get(static::INDEXER_ID);

-        /** @var Product\Action\Rows $action */
-        $action = $this->rowsActionFactory->create();
         if ($indexer->isScheduled()) {
-            $action->execute($ids, true);
+            $this->indexMutex->execute(
+                static::INDEXER_ID,
+                function () use ($ids) {
+                    $this->rowsActionFactory->create()->execute($ids, true);
+                }
+            );
         } else {
-            $action->execute($ids);
+            $this->rowsActionFactory->create()->execute($ids);
         }

         return $this;
diff --git a/vendor/magento/module-catalog/Model/Indexer/Product/Category.php b/vendor/magento/module-catalog/Model/Indexer/Product/Category.php
index e50b16837ee..b5ecd1076f9 100644
--- a/vendor/magento/module-catalog/Model/Indexer/Product/Category.php
+++ b/vendor/magento/module-catalog/Model/Indexer/Product/Category.php
@@ -5,7 +5,11 @@
  */
 namespace Magento\Catalog\Model\Indexer\Product;

+use Magento\Framework\Indexer\IndexMutexInterface;
+
 /**
+ * Product Categories indexer
+ *
  * @api
  * @since 100.0.2
  */
@@ -20,13 +24,17 @@ class Category extends \Magento\Catalog\Model\Indexer\Category\Product
      * @param \Magento\Catalog\Model\Indexer\Category\Product\Action\FullFactory $fullActionFactory
      * @param Category\Action\RowsFactory $rowsActionFactory
      * @param \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry
+     * @param IndexMutexInterface|null $indexMutex
+     * phpcs:disable Generic.CodeAnalysis.UselessOverridingMethod.Found
      */
     public function __construct(
         \Magento\Catalog\Model\Indexer\Category\Product\Action\FullFactory $fullActionFactory,
         Category\Action\RowsFactory $rowsActionFactory,
-        \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry
+        \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry,
+        ?IndexMutexInterface $indexMutex = null
     ) {
-        parent::__construct($fullActionFactory, $rowsActionFactory, $indexerRegistry);
+        //phpcs:enable
+        parent::__construct($fullActionFactory, $rowsActionFactory, $indexerRegistry, $indexMutex);
     }

     /**
diff --git a/vendor/magento/module-indexer/Model/IndexMutex.php b/vendor/magento/module-indexer/Model/IndexMutex.php
new file mode 100644
index 00000000000..fd0cf3dc480
--- /dev/null
+++ b/vendor/magento/module-indexer/Model/IndexMutex.php
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Copyright Â© Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Indexer\Model;
+
+use Magento\Framework\Indexer\ConfigInterface;
+use Magento\Framework\Indexer\IndexMutexException;
+use Magento\Framework\Indexer\IndexMutexInterface;
+use Magento\Framework\Lock\LockManagerInterface;
+
+/**
+ * Intended to prevent race conditions between indexers using the same index table.
+ */
+class IndexMutex implements IndexMutexInterface
+{
+    private const LOCK_PREFIX = 'indexer_lock_';
+
+    private const LOCK_TIMEOUT = 60;
+
+    /**
+     * @var LockManagerInterface
+     */
+    private $lockManager;
+
+    /**
+     * @var ConfigInterface
+     */
+    private $config;
+
+    /**
+     * @var int
+     */
+    private $lockWaitTimeout;
+
+    /**
+     * @param LockManagerInterface $lockManager
+     * @param ConfigInterface $config
+     * @param int $lockWaitTimeout
+     */
+    public function __construct(
+        LockManagerInterface $lockManager,
+        ConfigInterface $config,
+        int $lockWaitTimeout = self::LOCK_TIMEOUT
+    ) {
+        $this->lockManager = $lockManager;
+        $this->lockWaitTimeout = $lockWaitTimeout;
+        $this->config = $config;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $indexerName, callable $callback): void
+    {
+        $lockName = $indexerName;
+        $indexerConfig = $this->config->getIndexer($indexerName);
+        if (isset($indexerConfig['shared_index'])) {
+            $lockName = $indexerConfig['shared_index'];
+        }
+
+        if ($this->lockManager->lock(self::LOCK_PREFIX . $lockName, $this->lockWaitTimeout)) {
+            try {
+                $callback();
+            } finally {
+                $this->lockManager->unlock(self::LOCK_PREFIX . $lockName);
+            }
+        } else {
+            throw new IndexMutexException($indexerName);
+        }
+    }
+}
diff --git a/vendor/magento/module-indexer/etc/di.xml b/vendor/magento/module-indexer/etc/di.xml
index 9496f29cb1d..d7996ee68a7 100644
--- a/vendor/magento/module-indexer/etc/di.xml
+++ b/vendor/magento/module-indexer/etc/di.xml
@@ -12,6 +12,7 @@
     <preference for="Magento\Framework\Indexer\IndexerInterface" type="Magento\Indexer\Model\Indexer\DependencyDecorator" />
     <preference for="Magento\Framework\Indexer\Table\StrategyInterface" type="Magento\Framework\Indexer\Table\Strategy" />
     <preference for="Magento\Framework\Indexer\StateInterface" type="Magento\Indexer\Model\Indexer\State" />
+    <preference for="Magento\Framework\Indexer\IndexMutexInterface" type="Magento\Indexer\Model\IndexMutex" />
     <type name="Magento\Framework\Indexer\Table\StrategyInterface" shared="false" />
     <type name="Magento\Indexer\Model\Indexer">
         <arguments>
diff --git a/vendor/magento/framework/Indexer/IndexMutexException.php b/vendor/magento/framework/Indexer/IndexMutexException.php
new file mode 100644
index 00000000000..5148bb2bafd
--- /dev/null
+++ b/vendor/magento/framework/Indexer/IndexMutexException.php
@@ -0,0 +1,24 @@
+<?php
+/**
+ * Copyright Â© Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Framework\Indexer;
+
+use RuntimeException;
+
+/**
+ * Exception thrown when index lock could not be acquired
+ */
+class IndexMutexException extends RuntimeException
+{
+    /**
+     * @param string $indexerName
+     */
+    public function __construct(string $indexerName)
+    {
+        parent::__construct('Could not acquire lock for index: ' . $indexerName);
+    }
+}
diff --git a/vendor/magento/framework/Indexer/IndexMutexInterface.php b/vendor/magento/framework/Indexer/IndexMutexInterface.php
new file mode 100644
index 00000000000..25642ea1d4a
--- /dev/null
+++ b/vendor/magento/framework/Indexer/IndexMutexInterface.php
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Copyright Â© Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Framework\Indexer;
+
+/**
+ * Intended to prevent race conditions between indexers using the same index table.
+ */
+interface IndexMutexInterface
+{
+    /**
+     * Acquires a lock for an indexer, executes callable and releases the lock after.
+     *
+     * @param string $indexerName
+     * @param callable $callback
+     * @throws IndexMutexException
+     */
+    public function execute(string $indexerName, callable $callback): void;
+}
```

**Comment by hostep** [comment]
Looks like a similar change made it in the codebase some months ago: https://github.com/magento/magento2/commit/f6b12868406678f0c23a3dc4f9eb06fd187bf306

I can already see this code in Magento 2.4.4-beta4, so I'd assume this will be included in 2.4.4 as well.

**Comment by igorwulff** [comment]
Awesome @hostep !
